src/main/java/com/yelp/nrtsearch/server/ingestion/AbstractIngestor.java

```diff
@@ -61,6 +66,59 @@ public long addDocuments(List<AddDocumentRequest> addDocRequests, String indexNa
         .runIndexingJob();
   }
 
+  /**
+   * Delete documents matching the specified queries.
+   *
+   * <p>This method deletes documents from the index that match any of the provided queries. It is
+   * designed to support deletion of documents by ID or other criteria during ingestion (e.g., for
+   * handling DELETE and UPDATE_BEFORE row kinds in change data capture scenarios).
+   *
+   * <p>The delete operation is executed directly against the IndexWriter, ensuring it is processed
+   * before any subsequent addDocuments calls. This ordering is critical for UPDATE operations where
+   * the old document must be deleted before the new version is added.
+   *
+   * @param queries list of queries to match documents for deletion
+   * @param indexName target index name
+   * @return sequence number of the delete operation
+   * @throws Exception if delete fails
+   */
+  @Override
+  public long deleteByQuery(List<Query> queries, String indexName) throws Exception {
+    verifyInitialized(indexName);
+    IndexState indexState = globalState.getIndexOrThrow(indexName);
+    ShardState shardState = indexState.getShard(0);
```

Comment:
Why always 0? Should this not be based on what shard is currently active?

Reply:
Nrtsearch works with only one lucene shard per cluster and parallelism is achieved using parallel segment search and by sharding across clusters.